File: soyeun/피로도.py
# 던전의 개수가 1이상 8이하이므로 가능한 순서를 모두 탐색
import itertools

def solution(k, dungeons):
    answer = -1
    dungeons_case = list(itertools.permutations(dungeons,len(dungeons)))
    for case in dungeons_case:
        tired = k
        cnt = 0
        for need, minus in case:
            if tired >= need:
                cnt += 1
                tired -= minus
        answer = max(answer,cnt)
    return answer

k = 80
dungeons = 	[[80,20],[50,40],[30,10]]
print(solution(k,dungeons))

File: miri/피로도.java
class Solution {
    public int max;

    public int solution(int k, int[][] dungeons) {
        int[] visited = new int[dungeons.length];
        explore(dungeons, visited, k, 0);

        return max;
    }

    public void explore(int[][] dungeons, int[] visited, int k, int cnt) {
        for(int i = 0; i < dungeons.length; i++) {
            if(visited[i] == 0 && k >= dungeons[i][0]){
                visited[i] = 1;
                explore(dungeons,visited,k-dungeons[i][1],cnt+1);
                visited[i] = 0;
            }
        }

        if(max < cnt) {
            max = cnt;
        }
    }
}